### PR TITLE
add version for plugins

### PR DESCRIPTION
**Pull request by cristi1979**

### CODE REVIEW

File: manifests/plugin/abrt.pp [hunk 1]
```diff
@@ -48,7 +52,7 @@
   $group_real = pick($group, $::foreman_proxy::user)
   validate_string($group_real)
 
-  foreman_proxy::plugin { 'abrt': } ->
+  foreman_proxy::plugin { 'abrt': version => $version } ->
```
Comment (domcleal) [inline manifests/plugin/abrt.pp]: Nitpick, please split onto a new line

File: manifests/plugin/pulp.pp [hunk 2]
```diff
@@ -6,12 +6,16 @@
...
+# $version::      plugin package version, it's passed to ensure parameter of package resource
+#                 can be set to specific version number, 'latest', 'present' etc.
+#
 # $enabled::      enables/disables the plugin
 #
 # $pulp_url::     pulp url to use
 #
 class foreman_proxy::plugin::pulp (
   $enabled  = $::foreman_proxy::plugin::pulp::params::enabled,
+  $version  = $::foreman_proxy::plugin::pulpt::params::version,
```
Comment (domcleal) [inline manifests/plugin/pulp.pp]: typo, "pulp"

**Comment by domcleal** [comment]
:+1:

File: manifests/plugin.pp [hunk 3]
```diff
@@ -1,6 +1,7 @@
 # Installs a smart proxy plugin package
 define foreman_proxy::plugin(
-  $package = "${foreman_proxy::plugin_prefix}${title}",
+  $version = 'installed',
+  $package = "${foreman_proxy::params::plugin_prefix}${title}",
```
Comment (ekohl) [inline manifests/plugin.pp]: Why is this change needed?

File: manifests/plugin/abrt/params.pp [hunk 4]
```diff
@@ -1,6 +1,7 @@
 # Default parameters for the Pulp smart proxy plugin
 class foreman_proxy::plugin::abrt::params {
   $enabled                 = true
+  $version                 = 'installed'
```
Comment (ekohl) [inline manifests/plugin/abrt/params.pp]: I wonder if we should default to ```undef``` so we actually use the default in ```manifests/plugin.pp```. If that default is changed to ```foreman_proxy::plugin_version``` it would allow users to easily switch from installed to latest for every plugin.

Comment (cristi1979) [inline manifests/plugin/abrt/params.pp]: undef will set :undef for version, so the version will not be actually undefined in roder to use the default

File: manifests/plugin.pp (hunk 3 above)
Comment (cristi1979) [inline manifests/plugin.pp]: currently plugins can't be updated or kept in a certain version

Comment (ekohl) [inline manifests/plugin.pp]: I meant ```$package```. You're changing it from ```"${foreman_proxy::plugin_prefix}${title}"``` to ````"${foreman_proxy::params::plugin_prefix}${title}"```` and I don't see the need for that.

Comment (cristi1979) [inline manifests/plugin.pp]: Sorry about that, I didn't noticed. I've made the changes on my side a few days ago and I did it, it seems, on an older code, without doing a sync first. Fixed now

File: manifests/plugin/abrt/params.pp (hunk 4 above)
Comment (domcleal) [inline manifests/plugin/abrt/params.pp]: @ekohl do you mean remove $version from abrt::params, add `plugin_version = 'installed'` to the main params.pp and in the class just do `class foreman_proxy::plugin::abrt($version = $foreman_proxy::plugin_version)`, to save re-defining it?

Comment (ekohl) [inline manifests/plugin/abrt/params.pp]: No, I was suggesting setting it ```$version``` in ```abrt::params``` to ```undef``` so puppet will fall back to whatever ```foreman_proxy::plugin``` has as default for ```$version```.

Comment (domcleal) [inline manifests/plugin/abrt/params.pp]: Interesting, that would work.  I wonder if it's not explicit enough though?  At least if it's defined here (or as I described) then you'd see the current value of the parameter with kafo..

Comment (ekohl) [inline manifests/plugin/abrt/params.pp]: That's a fair point. On the other hand we also set group to undef and that works.

Comment (cristi1979) [inline manifests/plugin/abrt/params.pp]: I'm a little lost with this. What's the point in keeping abrt::params::version, since it will never be used? Should plugin::abrt::version default to foreman_proxy::plugin::version?

Comment (ekohl) [inline manifests/plugin/abrt/params.pp]: I'm not sure there is a point in keering ```abrt::params::version```. Maybe kafo needs it? @domcleal any reason we have group in params as well?

Comment (domcleal) [inline manifests/plugin/abrt/params.pp]: I don't think it needs it any more, it might have done in early versions.

Comment (cristi1979) [inline manifests/plugin/abrt/params.pp]: Should I push with group defaulting to foreman_proxy::user?
And remove the check pick($group, $::foreman_proxy::user)?

Comment (ekohl) [inline manifests/plugin/abrt/params.pp]: IIRC the pick was still needed because kafo hasn't loaded ```foreman_proxy``` and thus can't determine the value.

Comment (cristi1979) [inline manifests/plugin/abrt/params.pp]: But, by default, it will always pick between undef and ::foreman_proxy::user, so it must have access to ::foreman_proxy::user?

Comment (ekohl) [inline manifests/plugin/abrt/params.pp]: Correct, but kafo won't actually execute the code so that's not a problem. See https://github.com/theforeman/puppet-foreman_proxy/pull/121 for the discussion.

**Comment by cristi1979** [comment]
I've removed $group from <plugin>::params and default it to $::foreman_proxy::user also.

The same for $version and default it to $::foreman_proxy::plugin::version

File: manifests/plugin/abrt.pp [hunk 5]
```diff
@@ -30,7 +33,8 @@
 #
 class foreman_proxy::plugin::abrt (
   $enabled                 = $::foreman_proxy::plugin::abrt::params::enabled,
-  $group                   = $::foreman_proxy::plugin::abrt::params::group,
+  $version                 = $::foreman_proxy::plugin::version,
+  $group                   = $::foreman_proxy::user,
```
Comment (ekohl) [inline manifests/plugin/abrt.pp]: I'd leave this as is because it breaks the installer.

File: manifests/plugin/abrt.pp [hunk 6]
```diff
@@ -30,7 +33,8 @@
 #
 class foreman_proxy::plugin::abrt (
   $enabled                 = $::foreman_proxy::plugin::abrt::params::enabled,
-  $group                   = $::foreman_proxy::plugin::abrt::params::group,
+  $version                 = $::foreman_proxy::plugin::version,
```
Comment (ekohl) [inline manifests/plugin/abrt.pp]: I think this should default to 'undef' because kafo can't handle this. Also because I don't think you can refer to variables on a define in puppet.

File: manifests/plugin/abrt.pp (hunk 5 above)
Comment (cristi1979) [inline manifests/plugin/abrt.pp]: This makes me cry now :). It seems a recent push in the master branch removed those:

    $group_real = pick($group, $::foreman_proxy::user)
    validate_string($group_real)

Should I add them again?

Comment (ekohl) [inline manifests/plugin/abrt.pp]: No, those were intentional and solve it in a much better way by using ```foreman_proxy::settings_file``` instead of doing the same everywhere.

File: manifests/plugin/abrt.pp (hunk 6 above)
Comment (cristi1979) [inline manifests/plugin/abrt.pp]: But what will be the default? Should I add a pick:

    $version_real = pick($version, 'installed')
    validate_string($version_real)

Comment (ekohl) [inline manifests/plugin/abrt.pp]: I think you can just literally set ```$version = undef```.

**Comment by cristi1979** [comment]
any issues?

**Comment by ekohl** [comment]
I think I'd call it ```foreman_proxy::plugin_version```, but otherwise that sounds like a good idea.

**Comment by cristi1979** [comment]
I need some help with rspecs. When I'm setting $version = $foreman_proxy::plugin_version, the tests fail because:

    Failure/Error: should contain_package('myplugin').with_ensure('installed')
    expected that the catalogue would contain Package[myplugin] with ensure set to "installed" but it is set to nil

**Comment by cristi1979** [comment]
travis fixed

File: manifests/plugin.pp [hunk 7]
```diff
@@ -1,6 +1,7 @@
 # Installs a smart proxy plugin package
 define foreman_proxy::plugin(
-  $package = "${foreman_proxy::plugin_prefix}${title}",
+  $version = $foreman_proxy::plugin_version,
+  $package = "${foreman_proxy::params::plugin_prefix}${title}",
```
Comment (ekohl) [inline manifests/plugin.pp]: Rebase gone wrong? The package shouldn't change.

Comment (cristi1979) [inline manifests/plugin.pp]: fixed

File: spec/defines/foreman_proxy_plugin_spec.rb [hunk 8]
```diff
@@ -22,18 +22,28 @@
 
   context 'with package parameter' do
     let :params do {
-      :package => 'myplugin',
-    } end
+        :package => 'myplugin',
```
Comment (ekohl) [inline spec/defines/foreman_proxy_plugin_spec.rb]: Nitpick: odd indenting.

File: spec/defines/foreman_proxy_plugin_spec.rb [hunk 9]
```diff
@@ -42,12 +52,12 @@
 
   context 'when handling underscores on Debian' do
     let :facts do {
-      :osfamily => 'Debian',
-    } end
+        :osfamily => 'Debian',
```
Comment (ekohl) [inline spec/defines/foreman_proxy_plugin_spec.rb]: Nitpick: odd indenting.

File: spec/defines/foreman_proxy_plugin_spec.rb [hunk 10]
```diff
@@ -42,12 +52,12 @@
...
+      } end
 
     let :params do {
-      :package => 'my_fun_plugin',
-    } end
+        :package => 'my_fun_plugin',
```
Comment (ekohl) [inline spec/defines/foreman_proxy_plugin_spec.rb]: Nitpick: odd indenting.

File: spec/defines/foreman_proxy_plugin_spec.rb [hunk 11]
```diff
@@ -3,18 +3,18 @@
...
-    let :pre_condition do
-      'include foreman_proxy'
-    end
+  let :pre_condition do
+    'include foreman_proxy'
+  end
 
-    let :facts do {
+  let :facts do {
       :concat_basedir         => '/nonexistant',
```
Comment (ekohl) [inline spec/defines/foreman_proxy_plugin_spec.rb]: Nitpick: odd indenting.

**Comment by ekohl** [comment]
Besides nitpicks I think it's good.

**Comment by cristi1979** [comment]
it was the auto formatting from eclipse that did the strange indent.

**Comment by domcleal** [comment]
Merged as 33054bc, thanks @cristi1979!
